userspace/libsinsp/filter/ast.cpp

```diff
@@ -65,6 +199,109 @@ void base_expr_visitor::visit(list_expr* e) { }
 
 void base_expr_visitor::visit(unary_check_expr* e) { }
 
+void string_visitor::visit_logical_op(const char *op, std::vector<expr*> &children)
+{
+	bool first = true;
+
+	m_str += "(";
+
+	for (auto &c : children)
+	{
+		if(!first)
+		{
+			m_str += " ";
+			m_str += op;
+			m_str += " ";
+		}
+		first = false;
+		c->accept(this);
+	}
+	m_str += ")";
+}
+
+void string_visitor::visit(and_expr* e)
+{
+	visit_logical_op("and", e->children);
+}
+
+void string_visitor::visit(or_expr* e)
+{
+	visit_logical_op("or", e->children);
+}
+
+void string_visitor::visit(not_expr* e)
+{
+	m_str += "not ";
+
+	e->child->accept(this);
+}
+
+void string_visitor::visit(value_expr* e)
+{
+	m_str += ast::escape_str(e->value);
```

Comment:
Note (nitpicking): macros (a.k.a. identifiers, in the grammar) are represented by `value_expr` nodes too but should not be escaped. You can tell the difference by maintaining a state boolean in the visitor, turning it true if you visited a `binary_check_expr` node before getting here (and turning it off everywhere else).

Reply:
Actually, I'm not sure how a parser would ever generate an ast for a macro reference containing spaces. Both `some macro and proc.name=nginx` and `\"some macro\" and proc.name=nginx` result in parse failures. But I'll still include the handling for it, in case the parser starts handling it.